### ISSUE TITLE
Slowness on public health datahub

### ISSUE THREAD

**Issue by dsconnell** [berkeley-dsep-infra/hubs#3969]
### Bug description

Several individuals from courses I support have reported frustration with how slow RStudio on the public health datahub is at certain times. The complain is simply that code takes an unreasonably long time (sometimes minutes) to run. I set up a form to try to collect some data on when it's happening. Here are the reported dates and times so far (pacific time):

11/3/2022 | 9:00:00 PM
11/5/2022 | 8:37:00 AM
11/6/2022 | 6:00:00 PM
11/7/2022 | 6:00:00 PM
11/12/2022 | 8:31:00 AM
11/13/2022 | 5:38:00 AM
11/13/2022 | 5:00:00 PM
11/20/0022 | 5:15:00 PM

The pattern I noticed is that people are mostly having issues on weekends (11/5, 11/6, 11/12, 11/13, and 11/20). However, our students tend to be professionals with day jobs, so they do more work on weekends. That could explain the trend but I'm not sure.

Anyway, do you think the public health datahub simply needs more resources on weekends or is there some other issue?

### Environment & setup

- Hub: Public Health
- Language: R

### How to reproduce

1. Go to https://publichealth.datahub.berkeley.edu/
2. Run R code in RStudio
3. Wait a long time for results

**Comment by balajialg** [comment]
@dsconnell Yes - on first look, it seems there have been a few nonrunning user notebooks during the past week. We might need to provide more resources for the public health hub if the assumption is correct.
 
We will dig deeper into this once we are back from the holidays. For anything urgent this weekend, I recommend that your users use  [R Datahub](https://r.datahub.berkeley.edu/hub/login?next=%2Fhub%2F) if they have issues with the public health hub.

![image](https://user-images.githubusercontent.com/2306166/204383266-2263c88a-fb4e-4a74-bac3-edb63a35466d.png)

![image](https://user-images.githubusercontent.com/2306166/204383222-5c6c0951-4001-47c8-83d0-bf2c2b875fd3.png)


<img width="908" alt="image" src="https://user-images.githubusercontent.com/2306166/204055202-0e79e99f-9ebb-4198-a54c-44ea6313b133.png">

**Comment by balajialg** [comment]
@dsconnell Did you get any reports from students during the past 3-4 days?

**Comment by dsconnell** [comment]
@balajialg I have not received any reports during that time period! Did something change?

**Comment by balajialg** [comment]
@dsconnell Nothing changed. I was curious whether the hub posed any challenges for the students to work during the weekend.

**Comment by balajialg** [comment]
@dsconnell How are the hubs performing this semester? Are you observing same behavior currently? If yes - please do post more details in the comment or else close this issue. Thanks!

**Comment by dsconnell** [comment]
@balajialg I'm not working on any courses that chose to use the public health datahub this semester. Presumably courses are going to use it again in the fall, but frustrations with datahub were common on student evaluations, so the instructors might choose another path. I would love to tell them the public health datahub has increased resources and therefore is more reliable! Is that going to happen?

**Comment by balajialg** [comment]
@dsconnell Thanks for sharing the context! Can totally understand how frustrating it was for instructors and students last semester. Based on last semester's feedback, We made major upgrades to the infrastructure during the winter break which resulted in a reliable experience for instructors and students during the last 2-3 weeks (Making this assumption based on the feedback we received). So - I am expecting the courses using Public Health hub to have a similar experience this semester and would urge you to share this message with any instructors who are tentative about using Datahub.

**Comment by shaneknapp** [comment]
@dsconnell we have dedicated GCP, kubernetes and google filestore resources explicitly to nearly all hubs, including public health.  these dedicated resources mean that other hubs are in their own resource pool, and are independent and not impacted when other hubs have issues.  

since the beginning of the semester, we (knock on wood) have had no instances of hubs going down, and are currently averaging well over 4k users/day.

like @balajialg i believe that your experience will be *much* different this year.  i'd be happy to discuss this some more, show you pretty graphs, and hope that we can have you back in the family soon!

**Comment by balajialg** [comment]
Thanks, @shaneknapp for chiming in! 

@dsconnell Closing this github issue for now but please feel free to redirect any instructors our way if they express reservations. Thanks!

**Comment by dsconnell** [comment]
It's me again, sorry. An instructor for PHW251 (user ******) reported getting kicked out of the public health datahub several times on Sunday, August 27. Any idea why this would be happening?

(edited by shane to remove username)

**Comment by shaneknapp** [comment]
was this person actively using the hub when this happened (ie: typing/working/etc and kicked out), or some other behavior?

from looking at the logs, i see three times that their singleuser server was shut down due to inactivity at 7:45pm, 8:55pm and 10:43pm.  example:

```
(base) ➜  ~ grep -A 1 -B 3 "shutting" shutdown-server-debug.txt
[I 2023-08-27 23:18:06.111 SingleUserNotebookApp mixins:547] Starting jupyterhub single-user server version 4.0.2
[I 2023-08-27 22:43:03.597 SingleUserNotebookApp notebookapp:2179] Shutting down 0 terminals
[I 2023-08-27 22:43:03.597 SingleUserNotebookApp notebookapp:2164] Shutting down 0 kernels
[I 2023-08-27 22:43:03.596 SingleUserNotebookApp notebookapp:2067] No kernels or terminals for 1827 seconds; shutting down.
[I 2023-08-27 22:12:36.248 SingleUserNotebookApp log:191] 200 POST /user/wwheeler1/rstudio/events/get_events (*****@10.0.0.99) 8955.07ms
```

the other two instances of a shutdown that i found in the logs are identical except for the timestamp.

it looks to me that person was idle for 30m, and the session was terminated (see the timestamps on the bottom two entries in the snippet).  then they tried to resume their session, and a new server launches (timestamp at the top of the snippet).

this is expected behavior unless they were actively using the hub and were "kicked out".  if that's the case i'd need a better description about both what exactly they were doing and when this happened.
